File: Python/algo_codes/solu_py/array/traverse_array.py
class Solutions:
    
    def findMaxConsecutiveOnes(self, nums: list[int]) -> int:
        _max = count = 0
        for i in nums:
            count = (count + i) * i  # 遇0得0,遇1加1
            if count > _max:
                _max = count
        return _max

    def findPoisonedDuration(self, timeSeries: list[int], duration: int) -> int:
        t = 0
        if len(timeSeries) == 1:
            return duration
        last = timeSeries[0]
        for fast in timeSeries[1:]:
            if fast - last > duration:
                t += duration
            else:
                t += fast - last
            last = fast
        return t + duration

    def thirdMax(self, nums: list[int]) -> int:
        nums = list(set(nums))
        nums.sort(reverse=True)
        return nums[0] if len(nums) < 3 else nums[2]

    def maximumProduct(self , nums :list[int])->int:
        nums.sort(reverse=True)
        k = nums[0]*nums[1]*nums[2]
        i = nums[0]*nums[-1]*nums[-2]
        return i if i>k else k
        

File: Cpp/algo_codes/solu_cpp/traverse_array.cc
#ifndef CPP_TRAVERSE_ARRAY
#define CPP_TRAVERSE_ARRAY

#include <iostream>
#include <set>
#include <vector>
using namespace std;

class Solution {
 public:
  // 485
  int findMaxConsecutiveOnes(vector<int>& nums) {
    int max = 0, count = 0;
    for (auto c : nums) {
      if (c) count++;
      max = max > count ? max : count;
      if (!c) count = 0;
    }
    return max;
  }

  // 495
  int findPoisonedDuration(vector<int>& timeSeries, int duration) {
    int time_sum = 0;
    int n = timeSeries.size();
    for (int i = 0; i < n - 1; i++) {
      if ((timeSeries[i + 1] - timeSeries[i]) > duration)
        time_sum += duration;
      else
        time_sum += timeSeries[i + 1] - timeSeries[i];
    }
    time_sum += duration;
    return time_sum;
  }

  // 414
  int thirdMax(vector<int>& nums) {
    long max = LONG_MIN, mid_max = max, min_max = max;
    for (long n : nums) {
      if (n > max) {
        min_max = mid_max;
        mid_max = max;
        max = n;
      } else if (max > n && n > mid_max) {
        min_max = mid_max;
        mid_max = n;
      } else if (mid_max > n && n > min_max) {
        min_max = n;
      }
    }
    return min_max == LONG_MIN ? max : min_max;
  }
  int _thirdMax(vector<int>& nums) {
    set<int> s;
    for (auto num : nums) {
      s.insert(num);
    }
    set<int>::iterator it = s.end();
    return s.size() < 3 ? *(--it) : *(--(--(--it)));
  }

  // 628 本题是414的继承
  int maximumProduct(vector<int>& nums) {
    long max = LONG_MIN, mid_max = max, min_max = max, min = INT_MAX,
         pre_min = min;
    for (long n : nums) {
      if (n >= max) {
        min_max = mid_max;
        mid_max = max;
        max = n;
      } else if (max > n && n >= mid_max) {
        min_max = mid_max;
        mid_max = n;
      } else if (mid_max > n && n >= min_max) {
        min_max = n;
      }
      if (n < min) {
        pre_min = min;
        min = n;
      } else if (n >= min && n < pre_min) {
        pre_min = n;
      }
    }
    return min * pre_min * max > max * mid_max * min_max
               ? min * pre_min * max
               : max * mid_max * min_max;
  }
};

#endif